File: typescript/packages/core/src/workspace/executor/builtins/declare/export.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { IOResult } from '../../../../io/types.ts'
import { PolicyDenied } from '../../../../policy/errors.ts'
import { VarAttr } from '../../../../shell/variable.ts'
import { setAttr } from '../../../session/state.ts'
import type { SessionState } from '../../../session/session.ts'
import { exportedNames } from '../../../session/state.ts'
import type { SessionView } from '../../../../ops/types.ts'
import { ExecutionNode } from '../../../types.ts'
import { readonlyRefusal, refusal, requireView } from '../shared.ts'
import { EXPORT_FLAGS, EXPORT_USAGE } from './constants.ts'
import {
  declareLine,
  identifierFailure,
  identifierRefusal,
  splitDeclFlags,
  storeStagedArrays,
} from './declare.ts'
import type { BuiltinCall, Result } from '../types.ts'
import { sessionView } from '../../../session/state.ts'

function exportLines(session: SessionState, flags: Set<string>): string[] {
  // The exported set, not every shell variable: `X=hello` is absent and
  // `export Y=world` is present, which is what bash prints. -f selects
  // shell functions; mirage tracks no export attribute on functions, so
  // that form lists nothing, as bash does with none exported.
  //
  // Rendering is `declareLine`'s, not a second spelling of it: GNU's
  // `export -p` prints the *whole* cluster, so a readonly exported
  // scalar is `declare -rx R="1"` and an exported array is
  // `declare -ax AR=([0]="a")`. Writing `declare -x` here by hand
  // printed neither, and rendered an exported array as a bare
  // `declare -x AR` because it looked the value up among the scalars.
  if (flags.has('f')) return []
  return exportedNames(session)
    .map((name) => declareLine(session, name))
    .filter((line): line is string => line !== null)
}

export async function handleExport(
  assignments: string[],
  session: SessionState,
  state: SessionView | null = null,
  arrays: { name: string; append: boolean; items: string[] }[] | null = null,
): Promise<Result> {
  const { flags, names, bad } = splitDeclFlags(assignments, EXPORT_FLAGS)
  if (bad !== null) {
    const err = new TextEncoder().encode(`bash: export: -${bad}: invalid option\n${EXPORT_USAGE}`)
    return [
      null,
      new IOResult({ exitCode: 2, stderr: err }),
      new ExecutionNode({ command: 'export', exitCode: 2, stderr: err }),
    ]
  }
  if (names.length === 0 && (arrays === null || arrays.length === 0)) {
    const lines = exportLines(session, flags)
    const out = new TextEncoder().encode(lines.length > 0 ? `${lines.join('\n')}\n` : '')
    return [out, new IOResult(), new ExecutionNode({ command: 'export', exitCode: 0 })]
  }
  // -f is accepted and marks nothing: mirage carries no export attribute
  // on functions. -n is the off direction, and applies to both spellings,
  // since `export -n K=v` assigns and unexports.
  const view = requireView(state)
  const on = !flags.has('n')
  if (arrays !== null && arrays.length > 0) {
    // `export ARR=(a b)` marks the array as surely as it marks a scalar:
    // GNU prints `declare -ax ARR=([0]="a" [1]="b")`.
    const refused = await storeStagedArrays(
      'export',
      session,
      view,
      arrays,
      VarAttr.Export,
      on,
      true,
    )
    if (refused !== null) return refused
  }
  const errors: string[] = []
  for (const assign of names) {
    const badName = identifierRefusal('export', assign)
    if (badName !== null) {
      errors.push(badName)
      continue
    }
    const eq = assign.indexOf('=')
    if (eq >= 0) {
      const key = assign.slice(0, eq)
      if (view.isReadonly(key)) return readonlyRefusal('export', key)
      try {
        await view.set(key, assign.slice(eq + 1))
      } catch (err) {
        if (err instanceof PolicyDenied) return refusal('export', err)
        throw err
      }
      setAttr(session, key, VarAttr.Export, on)
    } else {
      // The bare form writes no value, so it marks through the plane's
      // no-value door rather than inventing an empty string. On a name
      // that does not exist yet that leaves it *unset and exported*,
      // which is bash's own third state -- `export Z` prints
      // `declare -x Z` and stays out of `env` until something gives it a
      // value. Still gated: marking a hidden or policy-refused name is a
      // session write.
      try {
        await view.mark(assign, VarAttr.Export, on)
      } catch (err) {
        if (err instanceof PolicyDenied) return refusal('export', err)
        throw err
      }
    }
  }
  if (errors.length > 0) return identifierFailure('export', errors)
  return [null, new IOResult(), new ExecutionNode({ command: 'export', exitCode: 0 })]
}

/** The `export` arm. */
export async function exportBuiltin(call: BuiltinCall): Promise<Result> {
  return handleExport(
    [...call.argv.args],
    call.session,
    sessionView(call.session, call.registry.policies),
  )
}

File: python/mirage/workspace/executor/builtins/declare/export.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.io import IOResult
from mirage.io.types import ByteSource
from mirage.ops.types import SessionView
from mirage.policy import PolicyDenied
from mirage.shell.variable import VarAttr
from mirage.workspace.executor.builtins.declare.constants import (EXPORT_FLAGS,
                                                                  EXPORT_USAGE)
from mirage.workspace.executor.builtins.declare.declare import (
    declare_line, identifier_failure, identifier_refusal, split_decl_flags,
    store_staged_arrays)
from mirage.workspace.executor.builtins.shared import (readonly_refusal,
                                                       refusal, require_view)
from mirage.workspace.executor.builtins.types import BuiltinCall, Result
from mirage.workspace.session import SessionState
from mirage.workspace.session.state import (exported_names, session_view,
                                            set_attr)
from mirage.workspace.types import ExecutionNode


def _export_lines(session: SessionState, flags: set[str]) -> list[str]:
    """Build sorted declaration lines for every exported name.

    The exported set, not every shell variable: ``X=hello`` is absent
    and ``export Y=world`` is present, which is what bash prints. ``-f``
    selects shell functions instead of variables; mirage tracks no
    export attribute on functions, so that form lists nothing, as bash
    does with none exported.

    Rendering is ``declare_line``'s, not a second spelling of it: GNU's
    ``export -p`` prints the *whole* cluster, so a readonly exported
    scalar is ``declare -rx R="1"`` and an exported array is
    ``declare -ax AR=([0]="a")``. Writing ``declare -x`` here by hand
    printed neither, and rendered an exported array as a bare
    ``declare -x AR`` because it looked the value up among the scalars.

    Args:
        session (SessionState): shell session state.
        flags (set[str]): option letters the caller supplied.

    Returns:
        list[str]: one declaration line per exported name.
    """
    if "f" in flags:
        return []
    lines = [declare_line(session, name) for name in exported_names(session)]
    return [line for line in lines if line is not None]


async def handle_export(
    assignments: list[str],
    session: SessionState,
    state: SessionView | None = None,
    arrays: list[tuple[str, bool, list[str]]] | None = None,
) -> tuple[ByteSource | None, IOResult, ExecutionNode]:
    """Export names, or print them (``export -p`` / bare ``export``).

    With no name operands, prints every entry in ``session.env`` as
    ``declare -x NAME="value"`` (bash's ``-p`` form). Invalid option
    characters fail with status 2 and the GNU usage line. Writes go
    through the session view, so readonly refusal and the pre_session
    policy gate fire here exactly as for any other writer.
    """
    flags, names, bad = split_decl_flags(assignments, EXPORT_FLAGS)
    if bad is not None:
        err = (f"bash: export: -{bad}: invalid option\n"
               f"{EXPORT_USAGE}").encode()
        return None, IOResult(exit_code=2,
                              stderr=err), ExecutionNode(command="export",
                                                         exit_code=2,
                                                         stderr=err)
    # -p with names is ignored for display; bare / -p alone print.
    if not names and not arrays:
        lines = _export_lines(session, flags)
        out = (("\n".join(lines) + "\n") if lines else "").encode()
        return out, IOResult(), ExecutionNode(command="export", exit_code=0)
    # -f is accepted and marks nothing: mirage carries no export
    # attribute on functions. -n is the off direction, and applies to
    # both spellings, since `export -n K=v` assigns and unexports.
    view = require_view(state)
    on = "n" not in flags
    if arrays:
        # `export ARR=(a b)` marks the array as surely as it marks a
        # scalar: GNU prints `declare -ax ARR=([0]="a" [1]="b")`.
        refused = await store_staged_arrays("export",
                                            session,
                                            view,
                                            arrays,
                                            mark=VarAttr.EXPORT,
                                            on=on,
                                            fatal=True)
        if refused is not None:
            return refused
    errors: list[str] = []
    for assign in names:
        bad_name = identifier_refusal("export", assign)
        if bad_name is not None:
            errors.append(bad_name)
            continue
        if "=" in assign:
            key, _, val = assign.partition("=")
            if view.is_readonly(key):
                return readonly_refusal("export", key)
            try:
                await view.set(key, val)
            except PolicyDenied as exc:
                return refusal("export", exc)
            set_attr(session, key, VarAttr.EXPORT, on)
        else:
            # The bare form writes no value, so it marks through the
            # plane's no-value door rather than inventing an empty
            # string. On a name that does not exist yet that leaves it
            # *unset and exported*, which is bash's own third state --
            # `export Z` prints `declare -x Z` and stays out of `env`
            # until something gives it a value. Still gated: marking a
            # hidden or policy-refused name is a session write.
            try:
                await view.mark(assign, VarAttr.EXPORT, on)
            except PolicyDenied as exc:
                return refusal("export", exc)
    if errors:
        return identifier_failure("export", errors)
    return None, IOResult(), ExecutionNode(command="export", exit_code=0)


async def export_builtin(call: BuiltinCall) -> Result:
    """The ``export`` arm.

    Args:
        call (BuiltinCall): the invocation.
    """
    return await handle_export(
        list(call.argv.args), call.session,
        session_view(call.session, call.namespace.registry.policies))
